src/tools/adig.c

```diff
@@ -151,15 +155,88 @@ static void print_help(void)
     "              SOA, SRV, TXT, TLSA, URI, CAA, SVCB, HTTPS\n\n");
 }
 
+static int split_rcline(char *rcline, char **rcargv, int len) {
+  int   rcargc;
+  char *last = NULL;
+
+  for (rcargc = 1, rcargv[rcargc] = strtok_r(rcline, " \n", &last);
+    rcargc < len && rcargv[rcargc];
+    rcargv[++rcargc] = strtok_r(NULL, " \n", &last))
+  {
+    DEBUGF(fprintf(stderr, "split_rcline() argc: %d, argv: %s\n", rcargc, rcargv[rcargc]));
+  }
+
+  return (rcargc);
+}
+
 static ares_bool_t read_cmdline(int argc, const char * const *argv,
-                                adig_config_t *config)
+                                adig_config_t *config, ares_bool_t is_rcfile)
 {
-  ares_getopt_state_t state;
-  int                 c;
+  ares_getopt_state_t  state;
+  int                  c;
+  char                 configdir[256];
+  char                *configdir_xdg;
+  unsigned int         cdlen = 0;
+  unsigned int         cdsize = sizeof(configdir);
+  char                *homedir;
 
   ares_getopt_init(&state, argc, argv);
   state.opterr = 0;
 
+  if (!is_rcfile) {
+
+#if defined(WIN32)
+    cdlen = (unsigned int) snprintf(configdir, cdsize, "%s", getenv("%%APPDATA%"));
+
+#elif defined(__APPLE__)
+    homedir = getenv("HOME");
+    if (homedir != NULL) {
+      cdlen = (unsigned int) snprintf(configdir, cdsize, "%s/%s/%s/%s", homedir, "Library", "Application Support", "c-ares");
+    }
+
+#else
+    configdir_xdg = getenv("XDG_CONFIG_HOME");
+
+    if (configdir_xdg == NULL) {
+      homedir = getenv("HOME");
+      if (homedir != NULL) {
+        cdlen = (unsigned int) snprintf(configdir, cdsize, "%s/%s", homedir, ".config");
+      }
+    }
+    else {
+      cdlen = (unsigned int) snprintf(configdir, cdsize, "%s", configdir_xdg);
+    }
+
+#endif
+
+    DEBUGF(fprintf(stderr, "read_cmdline() configdir: %s\n", configdir));
+
+    if (cdlen != 0 && cdlen < cdsize) {
+      char            rcfile[256];
+      static FILE    *rcdata = NULL;
+      char            rcline[256];
+
+      unsigned int    rclen;
+      unsigned int    rcsize = sizeof(rcfile);
+      int             rcargc;
+      char           *rcargv[64];
+
+      rclen = (unsigned int) snprintf(rcfile, rcsize, "%s/adigrc", configdir);
+      if (rclen < rcsize) {
+        rcdata = fopen(rcfile, "r");
+
+        if (rcdata != NULL) {
+          while (fgets(rcline, sizeof(rcline), rcdata) != 0) {
+            rcargc = split_rcline(rcline, rcargv, 62);
```

Comment:
What happens if the file has binary/non-printable data?

Reply:
The help menu will be printed.